File: Assignment-3/1/binary_search.py
def bin_search(array, low, high, search_key):
    """
        Actual logic goes here
    """
    if high < low:
        return low
    """
        (low + high) / 2 => may be float
        > It is type casted to integer
    """
    mid = int((low + high) / 2)
    """
        * If the element is present, finally the element will be in the mid
    """
    if array[mid] == search_key:
        return mid
    # Search on right side
    elif array[mid] < search_key:
        return bin_search(array, mid + 1, high, search_key)
    # Search on left side
    else:
        return bin_search(array, low, mid - 1, search_key)

def binary_search(array, search_keys, n, k):
    """
        For each elements in the search keys array 
        Binary search is called.
    """
    temp_index = 0
    while temp_index < k:
        print(bin_search(array, 0, n - 1, search_keys[temp_index]) , " ", end = "")
        temp_index += 1

def main():
    """
        Program starts here
    """

    #print("Give n and n elements:")
    first_line = [int(x) for x in input().split()]
    
    #print("Give k and k search elements:")
    second_line = [int(x) for x in input().split()]

    # Separating
    n = first_line[0]
    
    # list[x:] => means it outputs the elements from index x
    array = first_line[1:]

    # checking 
    assert(n == len(array))

    # Same applies here.
    k = second_line[0]
    search_keys = second_line[1:]
    assert(k == len(search_keys))
    binary_search(array, search_keys, n, k)

if __name__ == "__main__":
    main()

File: Assignment-3/1/binary_search.cpp
#include<iostream>
#include<vector>

#define NOT_FOUND -1;

using namespace std;

int bin_search(vector<long long> array, int low, int high, long long key)
{
    // If not found return -1.
    if(high < low)
        return NOT_FOUND;

    /*
        (low + high) / 2 => may be float
        > It is type casted to integer
    */
    int mid = (int)((low + high) / 2);
    
    /*
        * If the element is present, finally the element will be in the mid
    */
    if(array[mid] == key)
        return mid;
    // Search on right side
    else if(array[mid] < key)
    {
        return bin_search(array, mid + 1, high, key);
    }
    // Search on left side
    else
    {
        return bin_search(array, low, mid - 1, key);
    }
}

void binary_search(vector<long long> list, vector<long long> search_keys, int n, int k)
{
    //vector<int> result_indices(k);

    // temp_index is the index for search keys.
    int temp_index = 0;
    
    while(temp_index < k)
        {
            /*
                For each elements in search_keys perform binary search.
            */
            cout<<bin_search(list, 0, n - 1,search_keys[temp_index])<<" ";
            //result_indices[temp_index] = bin_search(list, 0, n - 1,search_keys[temp_index]);
            temp_index++;
        }

    //return result_indices;
}

int main()
{
    int n = 0;
    int k = 0;

    //cout<<"Give n:";
    cin>>n;

    //cout<<"Give n elements:";
    vector<long long> list(n);
    for(int i = 0; i < n; i++)
        cin>>list[i];
    
    //cout<<"Give k elements:";
    cin>>k;

    //cout<<"Give k search elements:";
    vector<long long> search_keys(k);
    for(int i = 0; i < k; i++)
        cin>>search_keys[i];
    
    //cout<<"Output:";
    binary_search(list, search_keys, n, k);
    
    /*
    vector<int> result_indices = binary_search(list, search_keys, n, k);

    for(int i = 0; i < k; i++)
        cout<<result_indices[i]<<" ";
    */

    cout<<"\n"; 
}